src/service/builder.js

```diff
@@ -0,0 +1,236 @@
+
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.observer_ = new win.IntersectionObserver((e) => this.observed_(e), {
+      // Root bounds are not important, so we can use the `root:null` for a
+      // top-level window.
+      root: isIframed(win) ? win.document : null,
+      rootMargin: '250% 31.25%',
+      threshold: 0.001,
+    });
+
+    /** @private @const {!Map<!AmpElement, {asap: boolean, isIntersecting: boolean}>} */
+    this.targets_ = new Map();
+
+    /** @private {?Array<!AmpElement>} */
+    this.parsingTargets_ = [];
+
+    /** @private {boolean} */
+    ampdoc.whenReady().then(() => this.checkParsing_());
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() =>
+      this.docVisibilityChanged_()
+    );
+  }
+
+  /** @override */
+  dispose() {
+    this.observer_.disconnect();
+    this.targets_.clear();
+    if (this.visibilityUnlisten_) {
+      this.visibilityUnlisten_();
+      this.visibilityUnlisten_ = null;
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  scheduleAsap(target) {
+    this.targets_.set(target, {asap: true, isIntersecting: false});
+    this.waitParsing_(target);
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  schedule(target) {
+    if (this.targets_.has(target)) {
+      return;
+    }
+
+    if (target.deferredBuild()) {
+      this.targets_.set(target, {asap: false, isIntersecting: false});
+      this.observer_.observe(target);
+    } else {
+      this.targets_.set(target, {asap: false, isIntersecting: true});
+    }
+
+    this.waitParsing_(target);
+    this.signalScanReady_();
+  }
+
+  /**
+   * @param {!AmpElement} target
+   */
+  unschedule(target) {
+    if (!this.targets_.has(target)) {
+      return;
+    }
+
+    this.targets_.delete(target);
+
+    this.observer_.unobserve(target);
+
+    if (this.parsingTargets_) {
+      removeItem(this.parsingTargets_, target);
+      this.checkParsing_();
+    }
+  }
+
+  /** @private*/
+  signalScanReady_() {
+    if (!this.scheduledReady_) {
+      this.scheduledReady_ = true;
+      const {win} = this.ampdoc_;
+      win.setTimeout(() => {
+        // This signal mainly signifies that some of the elements have been
+        // discovered and scheduled.
+        this.ampdoc_.signals().signal(READY_SCAN_SIGNAL);
+      }, 50);
+    }
+  }
+
+  /** @private */
+  docVisibilityChanged_() {
+    const vs = this.ampdoc_.getVisibilityState();
+    if (vs != VisibilityState.PAUSED && vs != VisibilityState.INACTIVE) {
+      this.targets_.forEach((_, target) => this.maybeBuild_(target));
+    }
+  }
+
+  /**
+   * @param {!AmpElement} target
+   * @private
+   */
+  waitParsing_(target) {
+    const parsingTargets = this.parsingTargets_;
+    if (parsingTargets && !parsingTargets.includes(target)) {
+      parsingTargets.push(target);
+      this.checkParsing_();
+    } else {
+      this.maybeBuild_(target);
+    }
```

Comment:
I think this if statement is incorrect. If `!parsingTargets`, we should definitely go to `maybeBuild_(target)`, but if there is we need a different behavior. Currently, `parsingTargets.includes(target)` can still send us to `maybeBuild_(target)`. I think what's supposed to happen is just a `checkParsing_()` call?

```suggestion
    if (parsingTargets) {
      if (!parsingTargets.includes(target)) {
        parsingTargets.push(target);
      }
      this.checkParsing_();
    } else {
      this.maybeBuild_(target);
    }
```

Reply:
Yep. You're right.